### ISSUE TITLE
Ability to disable SMBIOS parsing on boot, SMBIOS not present in TCG event logs

### ISSUE THREAD

**Issue by polarina** [systemd/systemd#29560]
### Component

systemd

### Is your feature request related to a problem? Please describe

We remotely attest to the health of a booted system by quoting PCR registers 0, 2, 3, 4, 7, 8, 9, and 14, and comparing them to a whitelist of authorized values. The system's booted using GRUB, hence PCR8 and PCR9.

Systemd parses extra options from SMBIOS who, if under attacker control, can compromises the security of the system without affecting the aforementioned PCR measurements.

We have been unable to locate the SMBIOS in our TCG event logs, thus we cannot attest to its integrity.

### Describe the solution you'd like

Can systemd provide a switch to disable parsing SMBIOS on boot? Say, for example, a kernel command line argument (which we can attest to via PCR8 measurements from GRUB).

### Describe alternatives you've considered

The only alternative I see is to maintain custom patches on top of systemd to remove SMBIOS functionality.

### The systemd version you checked that didn't have the feature you are asking for

254

**Comment by poettering** [comment]
SMBIOS is supposed to be measured into PCR 1. Why is that not enough?

systemd uses smbios data as input for a bunch of different things, as does the Linux kernel. what you are trying to do doesn't really work hence, you have to include PCR 1 in your analysis.

**Comment by polarina** [comment]
@bluca we deploy a fleet of physical IoT devices where we are concerned with physical access. Not in the, let's freeze-the-RAM-with-liquid-nitrogen kind of attacks, but from a software perspective. Say, someone temporarily disabling secure boot, booting into UEFI Shell to perform malicious changes, then boot back into our OS where the system has been left in a compromised state we can't detect.

**Comment by bluca** [comment]
Sounds like the actual problem is that secure boot can be disabled at all in such an environment?

**Comment by polarina** [comment]
Is that something that can even be solved? One can always plug a keyboard to the device, reboot, spam F2 to enter firmware setup, then disable secure boot. That by itself is not fatal to us, since the event is clearly measured in PCR7.

**Comment by polarina** [comment]
> SMBIOS is supposed to be measured into PCR 1. Why is that not enough?

@poettering not all platforms measure SMBIOS. I got two from computers from distinct vendors on my desk, where nothing is measured into PCR1 before the separator (and only boot options follow the separator).

**Comment by bluca** [comment]
> Is that something that can even be solved? One can always plug a keyboard to the device, reboot, spam F2 to enter firmware setup, then disable secure boot. That by itself is not fatal to us, since the event is clearly measured in PCR7.

Yes I am pretty sure you can build firmwares that disallow any kind of local access, and if your threat model entails local physical access, that should probably be the first thing to look at

**Comment by poettering** [comment]
> @poettering not all platforms measure SMBIOS. I got two computers from distinct vendors on my desk, where nothing is measured into PCR1 before the separator (and only boot options follow the separator).

that's a frmware bug. you might be able to hack around this, by early on measuring smbios yourself somewhere, before the first component consumes it. I would assume not just the kernel and systemd consume smbios, but grub too, hence this measurement probably has to be added to grub (or before grub) before smbios is consumed for the first time.

**Comment by poettering** [comment]
grub has an smbios command, so it's pretty clear it consumes it:

https://www.gnu.org/software/grub/manual/grub/html_node/smbios.html

it would be kinda pointless to only make the last element in the chain not use smbios. You have to accept every component in the chain uses and if it is not measured, then you have to look into changing that.

**Comment by polarina** [comment]
> grub has an smbios command, so it's pretty clear it consumes it:

Correct. However, GRUB measures commands before executing them into PCR8. If GRUB is instructed to read (potentially malicious) SMBIOS data, we will know about it and can reject the TPM quote.

**Comment by poettering** [comment]
> Correct. However, GRUB measures commands before executing them into PCR8. If GRUB is instructed to read (potentially malicious) SMBIOS data, we will know about it and can reject the TPM quote.

are you sure that it doesn't consume smbios even before it executes any commands? Also, even if, Linux kernel consumes smbios. a lot of it. for example module loading is bound to smbios (for laptop drivers and such), and there's so much more.

**Comment by poettering** [comment]
I am kinda tempted to close this. As discussed, this all appears very unconvincing to me.
